src/script/microbit-interfacing/MicrobitSerial.ts

```diff
@@ -161,49 +80,47 @@ class MicrobitSerial implements MicrobitConnection {
               inputBehaviour.buttonChange(sensorData.buttonB, 'B');
             }
           }
-        });
-      }
-      // TODO: Is this necessary? Will the DAPJs read() function ever return done=true?
-      if (done) {
-        break;
-      }
+        }
+      });
+    });
+
+    // There is an issue where we cannot read data out from the micro:bit serial
+    // buffer until the buffer has been filled.
+    // As a workaround we can spam the micro:bit with handshake messages until
+    // enough responses have been queued in the buffer to fill it and the data
+    // starts to flow.
+    let attempts = 0;
+    while (
+      this.serialProtocolState == SerialProtocolState.AwaitingHandshakeResponse &&
+      attempts++ < 20
+    ) {
+      const handshakeCmd = protocol.generateCommand(protocol.CommandTypes.Handshake);
+      writeLine(`Sending handshake ${handshakeCmd.message}`);
+      await this.usb.serialWrite(handshakeCmd.message);
+      await new Promise(resolve => setTimeout(resolve, 100));
+    }
+    if (this.serialProtocolState === SerialProtocolState.AwaitingHandshakeResponse) {
+      throw new Error('Handshake not received');
     }
   }
 
   public listenForDisconnect(callback: (event: Event) => unknown): void {}
 
   public removeDisconnectListener(callback: (event: Event) => unknown): void {}
 
-  private async connect(): Promise<void> {
-    let serialPort = this.usb.serialPort;
-    await serialPort.open({ baudRate: this.baudRate });
-    writeLine(`Opened with baudRate: ${this.baudRate}`);
-    if (serialPort.readable && serialPort.writable) {
-      this.connected = true;
-      this.encoder.readable.pipeTo(serialPort.writable);
-      serialPort.readable.pipeTo(this.decoder.writable);
-    } else {
-      throw new Error('Serial port not readable or writable');
-    }
-  }
-
   public isConnected(): boolean {
-    return this.connected;
+    return this.usb.isSerialConnected();
   }
 
   public disconnect(): void {
-    this.connected = false;
+    // Weirdly this disconnects the CortexM...
```

Comment:
I want to come back to this. I don't think MicrobitUSB should extend CortexM and I doubt this disconnect is intentional/necessary.

Reply:
Is that a question about extending something a bit more high level like the `DAPLink` class, or that it shouldn't extend anything and use composition instead?

Reply:
Not convinced it should extend anything. But will review on a separate PR.